File: binding/src/libqclipboard.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qclipboard.html
pub const qclipboard = struct {

    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QClipboard_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qclipboard.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QClipboard_Clear(@ptrCast(self));
    }

    pub fn SupportsSelection(self: ?*anyopaque, ) bool {
        return qtc.QClipboard_SupportsSelection(@ptrCast(self));
    }

    pub fn SupportsFindBuffer(self: ?*anyopaque, ) bool {
        return qtc.QClipboard_SupportsFindBuffer(@ptrCast(self));
    }

    pub fn OwnsSelection(self: ?*anyopaque, ) bool {
        return qtc.QClipboard_OwnsSelection(@ptrCast(self));
    }

    pub fn OwnsClipboard(self: ?*anyopaque, ) bool {
        return qtc.QClipboard_OwnsClipboard(@ptrCast(self));
    }

    pub fn OwnsFindBuffer(self: ?*anyopaque, ) bool {
        return qtc.QClipboard_OwnsFindBuffer(@ptrCast(self));
    }

    pub fn Text(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QClipboard_Text(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qclipboard.Text: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Text2(self: ?*anyopaque, subtype: []const u8, allocator: std.mem.Allocator) []const u8 {
        const subtype_str = qtc.libqt_string{
    .len = subtype.len,
    .data = subtype.ptr,
};
const _str = qtc.QClipboard_Text2(@ptrCast(self), subtype_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qclipboard.Text2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetText(self: ?*anyopaque, param1: []const u8) void {
        const param1_str = qtc.libqt_string{
    .len = param1.len,
    .data = param1.ptr,
};
qtc.QClipboard_SetText(@ptrCast(self), param1_str);
    }

    pub fn MimeData(self: ?*anyopaque, ) QtC.QMimeData {
        return qtc.QClipboard_MimeData(@ptrCast(self));
    }

    pub fn SetMimeData(self: ?*anyopaque, data: ?*anyopaque) void {
        qtc.QClipboard_SetMimeData(@ptrCast(self), @ptrCast(data));
    }

    pub fn Image(self: ?*anyopaque, ) QtC.QImage {
        return qtc.QClipboard_Image(@ptrCast(self));
    }

    pub fn Pixmap(self: ?*anyopaque, ) QtC.QPixmap {
        return qtc.QClipboard_Pixmap(@ptrCast(self));
    }

    pub fn SetImage(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QClipboard_SetImage(@ptrCast(self), @ptrCast(param1));
    }

    pub fn SetPixmap(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QClipboard_SetPixmap(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Changed(self: ?*anyopaque, mode: i32) void {
        qtc.QClipboard_Changed(@ptrCast(self), @intCast(mode));
    }

    pub fn OnChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QClipboard_Connect_Changed(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn SelectionChanged(self: ?*anyopaque, ) void {
        qtc.QClipboard_SelectionChanged(@ptrCast(self));
    }

    pub fn OnSelectionChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QClipboard_Connect_SelectionChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn FindBufferChanged(self: ?*anyopaque, ) void {
        qtc.QClipboard_FindBufferChanged(@ptrCast(self));
    }

    pub fn OnFindBufferChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QClipboard_Connect_FindBufferChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn DataChanged(self: ?*anyopaque, ) void {
        qtc.QClipboard_DataChanged(@ptrCast(self));
    }

    pub fn OnDataChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QClipboard_Connect_DataChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QClipboard_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qclipboard.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QClipboard_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qclipboard.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Clear1(self: ?*anyopaque, mode: i32) void {
        qtc.QClipboard_Clear1(@ptrCast(self), @intCast(mode));
    }

    pub fn Text1(self: ?*anyopaque, mode: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QClipboard_Text1(@ptrCast(self), @intCast(mode));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qclipboard.Text1: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Text22(self: ?*anyopaque, subtype: []const u8, mode: i32, allocator: std.mem.Allocator) []const u8 {
        const subtype_str = qtc.libqt_string{
    .len = subtype.len,
    .data = subtype.ptr,
};
const _str = qtc.QClipboard_Text22(@ptrCast(self), subtype_str, @intCast(mode));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qclipboard.Text22: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetText2(self: ?*anyopaque, param1: []const u8, mode: i32) void {
        const param1_str = qtc.libqt_string{
    .len = param1.len,
    .data = param1.ptr,
};
qtc.QClipboard_SetText2(@ptrCast(self), param1_str, @intCast(mode));
    }

    pub fn MimeData1(self: ?*anyopaque, mode: i32) QtC.QMimeData {
        return qtc.QClipboard_MimeData1(@ptrCast(self), @intCast(mode));
    }

    pub fn SetMimeData2(self: ?*anyopaque, data: ?*anyopaque, mode: i32) void {
        qtc.QClipboard_SetMimeData2(@ptrCast(self), @ptrCast(data), @intCast(mode));
    }

    pub fn Image1(self: ?*anyopaque, mode: i32) QtC.QImage {
        return qtc.QClipboard_Image1(@ptrCast(self), @intCast(mode));
    }

    pub fn Pixmap1(self: ?*anyopaque, mode: i32) QtC.QPixmap {
        return qtc.QClipboard_Pixmap1(@ptrCast(self), @intCast(mode));
    }

    pub fn SetImage2(self: ?*anyopaque, param1: ?*anyopaque, mode: i32) void {
        qtc.QClipboard_SetImage2(@ptrCast(self), @ptrCast(param1), @intCast(mode));
    }

    pub fn SetPixmap2(self: ?*anyopaque, param1: ?*anyopaque, mode: i32) void {
        qtc.QClipboard_SetPixmap2(@ptrCast(self), @ptrCast(param1), @intCast(mode));
    }
};

/// https://doc.qt.io/qt-6/clipboard.html#types
pub const enums = struct {
    pub const Mode = enum {
        pub const Clipboard: i32 = 0;
        pub const Selection: i32 = 1;
        pub const FindBuffer: i32 = 2;
        pub const LastMode: i32 = 2;
    };

};

File: binding/src/libqclipboard.cpp
#include <QClipboard>
#include <QImage>
#include <QMimeData>
#include <QObject>
#include <QPixmap>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qclipboard.h>
#include "libqclipboard.h"
#include "libqclipboard.hxx"

libqt_string QClipboard_Tr(const char* s) {
	QString _ret = QClipboard::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QClipboard_Clear(QClipboard* self) {
	self->clear();
}

bool QClipboard_SupportsSelection(const QClipboard* self) {
	return self->supportsSelection();
}

bool QClipboard_SupportsFindBuffer(const QClipboard* self) {
	return self->supportsFindBuffer();
}

bool QClipboard_OwnsSelection(const QClipboard* self) {
	return self->ownsSelection();
}

bool QClipboard_OwnsClipboard(const QClipboard* self) {
	return self->ownsClipboard();
}

bool QClipboard_OwnsFindBuffer(const QClipboard* self) {
	return self->ownsFindBuffer();
}

libqt_string QClipboard_Text(const QClipboard* self) {
	QString _ret = self->text();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QClipboard_Text2(const QClipboard* self, libqt_string subtype) {
	QString _ret = self->text(QString::fromUtf8(subtype.data, subtype.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QClipboard_SetText(QClipboard* self, const libqt_string param1) {
	self->setText(QString::fromUtf8(param1.data, param1.len));
}

const QMimeData* QClipboard_MimeData(const QClipboard* self) {
	return self->mimeData();
}

void QClipboard_SetMimeData(QClipboard* self, QMimeData* data) {
	self->setMimeData(data);
}

QImage* QClipboard_Image(const QClipboard* self) {
	return new QImage(self->image());
}

QPixmap* QClipboard_Pixmap(const QClipboard* self) {
	return new QPixmap(self->pixmap());
}

void QClipboard_SetImage(QClipboard* self, const QImage* param1) {
	self->setImage(*param1);
}

void QClipboard_SetPixmap(QClipboard* self, const QPixmap* param1) {
	self->setPixmap(*param1);
}

void QClipboard_Changed(QClipboard* self, int mode) {
	self->changed(static_cast<QClipboard::Mode>(mode));
}

void QClipboard_Connect_Changed(QClipboard* self, intptr_t slot) {
    void (*slotFunc)(QClipboard*, int) = reinterpret_cast<void (*)(QClipboard*, int)>(slot);
    QClipboard::connect(self, &QClipboard::changed, [self, slotFunc](QClipboard::Mode mode) {
	slotFunc(self, mode);
    });
}

void QClipboard_SelectionChanged(QClipboard* self) {
	self->selectionChanged();
}

void QClipboard_Connect_SelectionChanged(QClipboard* self, intptr_t slot) {
    void (*slotFunc)(QClipboard*) = reinterpret_cast<void (*)(QClipboard*)>(slot);
    QClipboard::connect(self, &QClipboard::selectionChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QClipboard_FindBufferChanged(QClipboard* self) {
	self->findBufferChanged();
}

void QClipboard_Connect_FindBufferChanged(QClipboard* self, intptr_t slot) {
    void (*slotFunc)(QClipboard*) = reinterpret_cast<void (*)(QClipboard*)>(slot);
    QClipboard::connect(self, &QClipboard::findBufferChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QClipboard_DataChanged(QClipboard* self) {
	self->dataChanged();
}

void QClipboard_Connect_DataChanged(QClipboard* self, intptr_t slot) {
    void (*slotFunc)(QClipboard*) = reinterpret_cast<void (*)(QClipboard*)>(slot);
    QClipboard::connect(self, &QClipboard::dataChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

libqt_string QClipboard_Tr2(const char* s, const char* c) {
	QString _ret = QClipboard::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QClipboard_Tr3(const char* s, const char* c, int n) {
	QString _ret = QClipboard::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QClipboard_Clear1(QClipboard* self, int mode) {
	self->clear(static_cast<QLCDNumber::Mode>(mode));
}

libqt_string QClipboard_Text1(const QClipboard* self, int mode) {
	QString _ret = self->text(static_cast<QLCDNumber::Mode>(mode));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QClipboard_Text22(const QClipboard* self, libqt_string subtype, int mode) {
	QString _ret = self->text(QString::fromUtf8(subtype.data, subtype.len), static_cast<QLCDNumber::Mode>(mode));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QClipboard_SetText2(QClipboard* self, const libqt_string param1, int mode) {
	self->setText(QString::fromUtf8(param1.data, param1.len), static_cast<QLCDNumber::Mode>(mode));
}

const QMimeData* QClipboard_MimeData1(const QClipboard* self, int mode) {
	return self->mimeData(static_cast<QLCDNumber::Mode>(mode));
}

void QClipboard_SetMimeData2(QClipboard* self, QMimeData* data, int mode) {
	self->setMimeData(data, static_cast<QLCDNumber::Mode>(mode));
}

QImage* QClipboard_Image1(const QClipboard* self, int mode) {
	return new QImage(self->image(static_cast<QLCDNumber::Mode>(mode)));
}

QPixmap* QClipboard_Pixmap1(const QClipboard* self, int mode) {
	return new QPixmap(self->pixmap(static_cast<QLCDNumber::Mode>(mode)));
}

void QClipboard_SetImage2(QClipboard* self, const QImage* param1, int mode) {
	self->setImage(*param1, static_cast<QLCDNumber::Mode>(mode));
}

void QClipboard_SetPixmap2(QClipboard* self, const QPixmap* param1, int mode) {
	self->setPixmap(*param1, static_cast<QLCDNumber::Mode>(mode));
}

